### PR TITLE
feature(mutator specification): Create the generic mutator specification

### PR DESCRIPTION
**Pull request by nicojs**
Create a generic mutator specification as a seperate npm package in order to reuse between the javascript-mutator and typescript-mutator (and other mutators in the future).

### CODE REVIEW

File: packages/stryker-mutator-specification/src/ArrayLiteralMutatorSpec.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      expectMutation(`['val']`, '[]');
+    });
+
+    it('should not mutate array initializers (leave that for ArrayNewExpressionMutator)', () => {
+      expectMutation(`new Array()`);
+      expectMutation(`new Array(1, 2, 3)`);
+    });
+
+    it('should mutate empty array literals as a filled array', () => {
+      expectMutation('[]', '[[]]');
```
Comment (simondel) [inline packages/stryker-mutator-specification/src/ArrayLiteralMutatorSpec.ts]: Do we always want to support this mutation? Adding it to the spec project would encourage that

File: packages/stryker-mutator-specification/src/ArrowFunctionMutatorSpec.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+export default function ArrowFunctionMutatorSpec(name: string, expectMutation: ExpectMutation) {
+  describe('ArrowFunctionMutator', () => {
+
+    it('should have name "ArrowFunction"', () => {
+      expect(name).eq('ArrowFunction');
+    });
+
+    it('should mutate an anonymous function with an inline return', () => {
+      expectMutation('const b = () => 4;', 'const b = () => undefined;');
```
Comment (simondel) [inline packages/stryker-mutator-specification/src/ArrowFunctionMutatorSpec.ts]: Does every language that we (plan) to support with this spec have arrow functions?

File: packages/stryker-mutator-specification/src/BlockMutatorSpec.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    it('should have name "Block"', () => {
+      expect(name).eq('Block');
+    });
+
+    it('should mutate the block of a function into an empty block', () => {
+      expectMutation('function() { return 4; }', 'function() {}');
+    });
+
+    it('should mutate a single block', () => {
+      expectMutation('const a = 3; { const b = a; }', 'const a = 3; {}');
```
Comment (simondel) [inline packages/stryker-mutator-specification/src/BlockMutatorSpec.ts]: Does every language that we (plan) to support with this spec have the `const` keyword?

File: packages/stryker-mutator-specification/src/ConditionalExpressionMutatorSpec.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+export default function ConditionalExpressionMutatorSpec(name: string, expectMutation: ExpectMutation) {
+
+  describe('ConditionalExpressionMutator', () => {
+
+    it('should have name "ConditionalExpression"', () => {
+      expect(name).eq('ConditionalExpression');
+    });
+
+    it('should replace conditional expressions', () => {
+      expectMutation('a < 3? b : c', 'false? b : c');
```
Comment (simondel) [inline packages/stryker-mutator-specification/src/ConditionalExpressionMutatorSpec.ts]: I would prefer to add a space between the condition and the `?`

File: packages/stryker-mutator-specification/src/ForStatementMutatorSpec.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    it('should have name "ForStatement"', () => {
+      expect(name).eq('ForStatement');
+    });
+
+    it('should mutate the condition of a for statement', () => {
+      expectMutation('for(let i=0;i<10; i++) { console.log(); }', 'for(let i=0;false; i++) { console.log(); }');
+    });
+
+    it('should mutate the condition of a for statement without a condition', () => {
+      expectMutation('for(let i=0;; i++) { console.log(); }', `for (let i = 0; false; i++) {${os.EOL}    console.log();${os.EOL}}`);
```
Comment (simondel) [inline packages/stryker-mutator-specification/src/ForStatementMutatorSpec.ts]: We are formatting our code here. Is that intentional?  We don't format the previous for statement. How will we make sure that all mutators return this exact formatting change?

File: packages/stryker-mutator-specification/src/IfStatementMutatorSpec.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+export default function IfStatementMutatorSpec(name: string, expectMutation: ExpectMutation) {
+
+  describe('IfStatementMutator', () => {
+
+    it('should have name "IfStatement"', () => {
+      expect(name).eq('IfStatement');
+    });
+
+    it('should mutate an expression to `true` and `false`', () => {
+      expectMutation('if(something){ a++ }', 'if(true){ a++ }', 'if(false){ a++ }');
```
Comment (simondel) [inline packages/stryker-mutator-specification/src/IfStatementMutatorSpec.ts]: This code may be formatted by mutators by introducing a space between `if` and `(something)`

File: packages/stryker-mutator-specification/src/ArrayLiteralMutatorSpec.ts (hunk 1 above)
Comment (nicojs) [reply packages/stryker-mutator-specification/src/ArrayLiteralMutatorSpec.ts]: Maybe we should change `['Stryker was here!']`, would be more explicit.

File: packages/stryker-mutator-specification/src/ArrowFunctionMutatorSpec.ts (hunk 2 above)
Comment (nicojs) [reply packages/stryker-mutator-specification/src/ArrowFunctionMutatorSpec.ts]: Yes, the idea is that all "javascript like languages" should support mutation of arrow functions.

File: packages/stryker-mutator-specification/src/ConditionalExpressionMutatorSpec.ts (hunk 4 above)
Comment (nicojs) [reply packages/stryker-mutator-specification/src/ConditionalExpressionMutatorSpec.ts]: That shouldn't matter right? Formatting does not change the intent here.

File: packages/stryker-mutator-specification/src/ForStatementMutatorSpec.ts (hunk 5 above)
Comment (nicojs) [reply packages/stryker-mutator-specification/src/ForStatementMutatorSpec.ts]: Agreed, let's make the code here readable and change the expect implementation for typescript.

File: packages/stryker-mutator-specification/src/ArrowFunctionMutatorSpec.ts (hunk 2 above)
Comment (nicojs) [reply packages/stryker-mutator-specification/src/ArrowFunctionMutatorSpec.ts]: I'll add it in the readme 👍

File: packages/stryker-mutator-specification/src/BlockMutatorSpec.ts (hunk 3 above)
Comment (nicojs) [reply packages/stryker-mutator-specification/src/BlockMutatorSpec.ts]: Yep, JSX, javascript and TypeScript support it. If it becomes a problem we can change it later

File: packages/stryker-mutator-specification/src/IfStatementMutatorSpec.ts (hunk 6 above)
Comment (nicojs) [reply packages/stryker-mutator-specification/src/IfStatementMutatorSpec.ts]: Ok, i'll do that

**Comment by simondel** [comment]
Awesome!

**Pull request merged**